Design Doc for Web Components

bkjohnson opened the pull request:
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/18150

The doc can be viewed [here](https://github.com/department-of-veterans-affairs/va.gov-team/blob/wc-design-doc/platform/engineering/design-docs/2021-01-13-web-component-library.md).

bkjohnson committed:
Copy template

bkjohnson committed:
Draft the Overview

bkjohnson committed:
Begin adding detailed design, testing, questions

bkjohnson committed:
Remove some instructions

bkjohnson committed:
Add more detail - talk about alternatives & future

bkjohnson committed:
Fix formatting/whitespace up top

bkjohnson committed:
Add questions & risk

bkjohnson committed:
Add some initial work estimates

bkjohnson committed:
Fix typos, remove placeholders, add background bit

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+### Objective
+
+The goal is to establish a component library based on Web Components, which are a browser standard. This will supercede the current `formation-react` library and some of the liquid HTML templates currently in `vets-website`. A non-goal is creating an entirely new visual experience as part of the component library.
```
1. Fun fact (which I just learned): `supercede` is apparently a "disputed variant spelling" of supersede.
2. Additionally, we can probably replace `formation-react` with `component-library` now.
3. Is the goal to completely replace the existing component library with web components?

bkjohnson replied:
3. I would like that to be the goal.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+### Background
+
+We currently have some "components" implemented in HTML/CSS (and sometimes JS) in the `formation` npm package. These can be used anywhere where `formation` is loaded. We also have `formation-react`, which is a package with some of those patterns implemented as React components. These can only be used in React apps. Lastly, in `vets-website` we have some Liquid templates which implement some of th ...
```
For clarity, prefer bullet points. (Source: [Google Technical Writing One](https://developers.google.com/tech-writing/one/lists-and-tables) course.)

Maybe something like:
```markdown
The problem this design document intends to solve is the re-implementation of common components across different repositories and libraries, using different technologies.

Currently, common components may be found in one or more of three places:
- The `formation` NPM package
  - More information here
- The `component-library` NPM package
  - More information here
- Liquid templates in `vets-website`
  - More information here
```

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+The goal is to establish a single codebase and documentation for the implementation of design system components. This will allow the components to be used in React apps, static HTML, or any other web framework.
+
+The Web Component library will live in its own repo and will be added to `vets-website` as a dependency, where the components will be loaded and defined for the browser to use. As more components are added to the Web Component version of the library, the older React-specific components will be phased out.
```
> As more components are added to the Web Component version of the library, the older React-specific components will be phased out.

Tell me more about this in the detailed design. How? What's the migration process look like? When will the component be removed from the library?

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+### Detailed Design
+
+In the `component-library` repo, we will start a fresh orphan branch for the Web Component version of the library to live. If we reach a point where the React version of the library is no longer in use, we will rebase the orphan Web Component branch off of master (in order to cleanly preserve the history) and that will become the new default branch going forward.
```
Have you tried doing this to make sure it'll work like you want? I'm not super familiar with rebasing.

When you say "orphan branch," the web component branch will share a history with `master`, right?

bkjohnson replied:
I left out some detail here. I have done all of this locally except for the rebasing, but I'm pretty confident it will work like I expect. I created a branch off of `master` called `wiped` where I deleted everything but the `.git` directory, and then I created the orphan branch off of that. That gave me a branch with a clean slate in the same repo, but none of the history.

---

I just did the rebase as a sanity check, and the only conflicts were `.gitignore`, `package.json`, and the `.storybook` `main.js` and `preview.js` files, which is what I expected.

The result is [this branch](https://github.com/department-of-veterans-affairs/component-library/tree/wc-pilot-rebase-test). The commits are clean, but I have some more things to delete which isn't a big deal.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+In the `component-library` repo, we will start a fresh orphan branch for the Web Component version of the library to live. If we reach a point where the React version of the library is no longer in use, we will rebase the orphan Web Component branch off of master (in order to cleanly preserve the history) and that will become the new default branch going forward.
+
+Until we reach that point, the WC branch will be developed in the `component-library` repo alongside the main branch containing the React components. The WC branch will not be published to npm, and the `vets-website` dependency will instead use a Git url with a tag.
```
Tell me about this tag. When will it be created?

Link to the NPM documentation that says this is possible.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+## Specifics
+
+### Detailed Design
```
I recommend breaking this up into subsections. Maybe the following:
- Migration to web components
- Tooling
- Loading web components on VA.gov
- Styling

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+One of the additional benefits of Web Components is how they can be styled. Each Component can have its own styles scoped to its [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) (this means that if a Component renders a `<div>`, the CSS can broadly style the `div` node without worrying about it affecting all the other `<div>`s on the page). As a result, e ...
+
+Once in `vets-website`, the setup code for the WC library will be in its own entry file that is loaded on each page. In addition, this script will be one of the first loaded in order to ensure that the Components will be defined in the browser before they are rendered on the page.
```
- What does the "setup code for the WC library" do?
- How will the script be one of the first loaded?
- Tell me about lazy loading

Don't forget to link to the appropriate Stencil docs so you don't have to write it all yourself. :wink:

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+Accessibility and performance are two things that should receive lots of attention.
+
+Do we want to remove the explicit dependency on [USWDS](https://designsystem.digital.gov/)? Rather than explicitly relying on `uswds`, the Component library could be an abstraction for `vets-website` that implements the appropriate styles from `uswds`. A Component library complete with global styling defaults might be able to accomplish everything that `uswds` is providing.
```
Two things come to mind:
1. Because we can independently style web components, is it feasible to adopt USWDS 2.0 styles in just our web components?
1. I believe the design system's component library should be an implementation of the visual design language that the design system defines. It should be opinionated about its styles and provide overrides only as needed. (To maintain consistency across all of VA.gov, those needs should be deliberate—more than just, "I don't like it.")

bkjohnson replied:
1. The short answer is "probably". The longer answer would involve looking at the differences between USWDS 1.0 and 2.0 and seeing where they differ. However, we can make the CSS for each component be whatever we want. This could mean copy-pasting USWDS stuff (not ideal), or adding it as a dependency and `@import`ing the appropriate CSS files into each component and then making any additional modifications that we want.

With that being said, I think that adopting USWDS 2.0 as part of this design document would fall under non-goals:

> A non-goal is creating an entirely new visual experience as part of the component library.

2. I agree.

cvalarida replied:
My opinion: Let's say we're going to use USWDS 2.0 for our WC styling and if we run into any problems with that, we can reassess later. The design document is meant to be updated if we find the implementation needs to change.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,139 @@
+
+#### Risks
+
+Despite polyfills, there is some risk of certain Web Component features not working properly in all browsers (IE 11, older versions of Edge). Due to existing polyfill support combined with the trend of users switching to more modern browsers over time, this is a low risk. If we encounter this problem, the workaround will be to avoid using whatever feature is under consideration and instead implem ...
```
```suggestion
Despite polyfills, there is some risk of certain Web Component features not working properly in all browsers (IE 11, older versions of Edge). Due to existing polyfill support combined with the trend of users switching to more modern browsers over time, this is a low risk. To mitigate this risk, we will only use Web Component features which are supported with or without pollyfills on all [supported browsers](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/platform/front-end-standards/documented-decisions/browser-support/).
```

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```suggestion
There isn't another way of achieving a proper (i.e., not copy-pasted HTML) component library that isn't specific to a single framework like React. However, there are other tools/frameworks besides Stencil that we could use to build Web Components.
```

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
This should be called out in the detailed design.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```suggestion
```

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```suggestion
Continue adding Components until the React version of `component-library` is completely replaced. Web Components can co-exist fine with React components, but it will be easier if we only have to maintain a single library.
```

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
Maybe link to https://stenciljs.com/docs/static-site-generation?

I'd also recommend saying only that this is an _option_, not a thing we will definitely do as future work.

cvalarida reviewed (commented):
This is a good first draft, but I'd like to see more meat in the detailed design section.

Design documents are intended to be very technical. Feel free to get into the weeds with the implementation details! Tell me how we load the WC loader. Tell me how that loader works. Tell me about lazy loading the components. Tell me about the performance implications and how Stencil mitigates that risk. Tell me why I might be nervous about adding yet another technology to the stack. Tell me how our implementation addresses my concerns. Tell me about the relationship between React and web components. Can I treat them like any other blackbox React component?

I know we've talked about these things in Slack, so I know you've thought about them, but they should be captured here as well. Be sure to heavily reference the Stencil documentation.

bkjohnson committed:
Apply suggestions from code review

Co-authored-by: Chris Valarida <chris.valarida@adhocteam.us>

bkjohnson committed:
Replace formation-react

bkjohnson committed:
Add detail on tagging

bkjohnson committed:
Break detailed design up into groups

bkjohnson committed:
Break background up into bullets

bkjohnson committed:
Add detail on phasing out React components

bkjohnson committed:
Update revision history

I'll still need to update it once the doc is approved

bkjohnson committed:
Add Typescript to Tooling section

bkjohnson committed:
Add more info to Templating as future work

bkjohnson committed:
Add more detail on loading the Components

bkjohnson committed:
Clarity tweaks

bkjohnson committed:
Polish things a bit

bkjohnson committed:
Break questions up

bkjohnson committed:
Add a day to each estimate

bkjohnson committed:
Update whitespace for proper indentation

bkjohnson committed:
Add Tooling info & a caveat

bkjohnson committed:
Add future work on performance improvements

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+#### Loading Web Components on va.gov
+
+Once in `vets-website`, the setup code for the WC library will be in its own entry file that is loaded on each page. This code will be responsible for applying polyfills for older browsers as well as defining the Web Components.
```
I think this is what you mean?
```suggestion
The setup code for the WC library will be in its own `vets-website` entry file that is loaded on each page. This code will be responsible for [applying polyfills](https://stenciljs.com/docs/faq#what-polyfills-does-stencil-provide) for older browsers as well as [defining the Web Components](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define).
```
There may be better links for both of those. :man_shrugging:

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+In addition, this script will be one of the first loaded in order to ensure that the Components will be defined in the browser before they are rendered on the page. This will be done by adding a `<script>` entry towards the top of [the `<head>` tag that gets loaded on every page](https://github.com/department-of-veterans-affairs/vets-website/blob/dc5018065eccf073d172097a19922755126c98a3/src/site/ ...
+
+For further optimization, the Components will be lazy-loaded. If our Web Component library ends up having hundreds of Components, this could cause a significant performance impact if we defined these every time a new user landed on a va.gov page. Instead, the Components will only _actually_ be defined if they appear on the page. [Stencil has a useful article that talks about their lazy loading ap ...
```
```suggestion
For further optimization, the Components will be lazy-loaded. Even if our Web Component library ends up having hundreds of Components, lazy loading will reduce the risk of a significant negative performance impact. The Components will only _actually_ be defined if they appear on the page. [Stencil has a useful article that talks about their lazy loading approach](https://stenciljs.com/blog/how-lazy-loading-web-components-work).
```

> The Components will only _actually_ be defined if they appear on the page.

Is this true? My understanding from that article is that it registers _wrappers_ for **all** the components, and then when they're rendered by the browser, the component wrapper renders the "skeleton" (not a real term) and makes the async call to fetch the actual component logic.

We might reference Ionic as an example of a really popular Web Component library using Stencil which has 100+ components. Bonus points for performance metrics.

bkjohnson replied:
Yeah, I glossed over the wrapper part. I'll be a bit more specific.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+Each Component will have:
+
+- Unit and E2E tests written using Jest
```
And whatever Stencil gives us? https://stenciljs.com/docs/testing-overview

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+N/A
+
+### Debugging
```
How will Sentry fit into this? (Maybe the answer is that it won't, but it's worth calling out explicitly.)

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+### Caveats
+
+Even though Web Components are already a feature provided by the browser, this is a feature that the VA Design System hasn't yet taken advantage of. By introducing this "new" technology, there is a possibility of this just turning into technical debt if it exists alongside the current `component-library`, `formation` component styles, and Liquid templates for components. In order to avoid this, w ...
```
From the template, this section is for
> Gotchas, differences between the design and implementation, other potential stumbling blocks for users or maintainers, and their implications and workarounds. Unless something is known to be tricky ahead of time, this section will probably start out empty.

What you've got here seems more like a risk to me. What do you think?

bkjohnson replied:
Or right! I meant to leave a comment on this since I had a question about this exact thing :grin: . I put it in Caveats because:

> other potential stumbling blocks for users or maintainers, and their implications

I'll go ahead and move it to Risks though.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
Bonus points for mentioning how cache busting will work.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
I'm not recommending or asking for a change here right now since this is in `Future Work`, but food for thought: Are you saying the `MegaMenu` should be a component in the component library, then? If not, will there be two separate WC loaders—one from `component-library` and one from `vets-website`? Or is there a way we can use the one loader and register components from both places?

bkjohnson replied:
I'm not saying we _should_ turn `MegaMenu` into a Web Component. There's a lot going on there and we'd probably have to break things down a ~~bit~~ lot. I'm just pointing out that on static pages, `MegaMenu` is the only reason why React is used at all. If we eliminated that, we could improve performance overall.

cvalarida replied:
I think the log in / search component up at the top is also built using React and is not part of the `MegaMenu`. Mmm...I _think_ it's not a part of it, anyhow. Turns out I'm too lazy to look right now, though. :see_no_evil:

![image](https://user-images.githubusercontent.com/12970166/106044191-38d76a80-6094-11eb-9920-eaa735f04008.png)

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+- [Liquid templates in `vets-website`](https://github.com/department-of-veterans-affairs/vets-website/tree/958d89f18db29a1f3e1f1cbe4e304299eb387de7/src/site)
+    - This is similar to the static markup found in `formation`. These templates of components are only used in the content build for va.gov
+
+Maintaining multiple sources of truth is a burden for all teams involved and leads to wasted time.
```
Tell me how using Web Components is going to solve this problem.

bkjohnson replied:
From the design doc template:

> The background section should contain information the reader needs to know to understand the problem being solved. This can be a combination of text and links to other documents.
>
> Do NOT describe the solution here. That goes in High Level Design.

Web Components are part of the solution, so should I really mention that here?

cvalarida replied:
No, you're right. I was looking for the first sentence of the `High Level Design` you've already got. (Literally the sentence after this one. :facepalm:) I completely skipped over it.

>Maintaining multiple sources of truth is a burden...

>The goal is to establish a single codebase and documentation for the implementation of design system components.

You've got it covered. :+1: 

Bonus points for using the same terminology, though. Maybe something like

> Maintaining multiple codebases and sources of documentation for effectively the same components is a burden...

cvalarida reviewed (commented):
I left a bunch of comments, but I hope it's less work than it looks! :sweat_smile: They're categorized roughly like so:
- High priority
  - The migration plan—this is the only one that's really stopping me from give this my stamp of approval right now
- Medium priority
  - Minor questions like in the `Testing Plan` and `Debugging` sections
- Low priority
  - Edits for grammar and clarity

bkjohnson committed:
Apply some low-priority edits

Co-authored-by: Chris Valarida <chris.valarida@adhocteam.us>

bkjohnson commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+> `yarn add <git remote url>#<branch/commit/tag>` installs a package from a remote git repository at specific git branch, git commit or git tag.
+
+We will use a [Github Release](https://github.com/department-of-veterans-affairs/component-library/releases/new) to create a new tag for the branch we want. The tags will follow [semantic versioning](https://semver.org/), and will be created after enough new development has landed in the branch (i.e. a new component, bugfix, etc. Documentation typo fixes won't get a release). A tag will have the  ...
```
I'll update `a` to `A`, but I think the period should remain after the parenthesis.

bkjohnson committed:
Apply some more edits

bkjohnson committed:
Update platform/engineering/design-docs/2021-01-13-web-component-library.md

Co-authored-by: Chris Valarida <chris.valarida@adhocteam.us>

bkjohnson committed:
Mention cache busting for the sweet bonus points

bkjohnson committed:
Mention explicit USWDS dependency as future work

bkjohnson committed:
Merge branch 'wc-design-doc' of github.com:department-of-veterans-affairs/va.gov-team into wc-design-doc

bkjohnson committed:
Be more specific about lazy-loading

bkjohnson committed:
Change the migration plan

bkjohnson committed:
Reduce an indent

bkjohnson committed:
Reorder some sections in Detailed Design

bkjohnson committed:
Mention Chromatic as possible future work

bkjohnson commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,185 @@
+
+We will use a [Github Release](https://github.com/department-of-veterans-affairs/component-library/releases/new) to create a new tag for the branch we want. The tags will follow [semantic versioning](https://semver.org/), and will be created after enough new development has landed in the branch (i.e. a new component, bugfix, etc. Documentation typo fixes won't get a release). A tag will have the  ...
+
+If a new Web Component is created that replaces an existing React component, the Web Component dependency in `vets-website` will be updated to the new tag and VFS teams will be able to use this component in their code according the the associated documentation for how to use the component. After a sprint, the React component will be removed from `component-library` and a new npm version will be p ...
```
I've made some updates to the migration plan which follow a less aggressive approach! This is based off of some learnings from the following articles:

- https://lethain.com/migrations/
- https://increment.com/cloud/case-studies-in-cloud-migration/ (specifically the section on Pinterest)

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+One of the additional benefits of Web Components is how they can be styled. Each Component can have its own styles scoped to its [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) (this means that if a Component renders a `<div>`, the Component's CSS can broadly style the `div` node without worrying about it affecting all the other `<div>`s on the page). As ...
+
+For now each Component will have its own styles _implicitly_ based off of USWDS, which resemble what is currently available in the Design System. Potential future work would involve explicitly importing USWDS component stylesheets into each Component's CSS file.
```
:+1: 

Bonus points for citing the experiment you did to use USWDS 2.0 in here and the trouble you ran into. This experiment informed your decision, so it's worth calling out that you _did_ try. In case somebody asks later. :wink:

bkjohnson replied:
I'll link to the section in Future Work where I talk about this.

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+#### Migration to Web Components
+
+Once a new release is made for the Web Component library, we will [configure a linting rule](https://github.com/AlexMost/eslint-plugin-deprecate/blob/master/docs/rules/import.md) as part of the `lint:js:changed` script in `vets-website` to throw an error if a React component is imported that has a released Web Component replacement. This will prevent the older React component from appearing in mo ...
```
:thinking: _Can_ we add this linting rule only for the `lint:js:changed` NPM script?

With this process, if an engineer makes _any_ changes to a file that imports the legacy component, they'll be forced to update the component usage before committing (or ignore the commit hook linting), right?

For the most part, forcing an update is probably fine if we're not changing the API of the component. If it's just like a `<TextInput>` -> `<va-text-input>` rename, great, but if it'll require more involved changes, that's probably not awesome.

How might we reduce the burden for engineers making unrelated changes? Is there any way to catch only _new_ imports for legacy components?

bkjohnson replied:
> _Can_ we add this linting rule only for the lint:js:changed NPM script?

Yes, but that will likely involve some minor refactoring of the current config and/or the script itself. One option would be to change [the current command](https://github.com/department-of-veterans-affairs/vets-website/blob/9da6a50293000210257703ffef135c10fb6fe6d9/package.json#L38) from `eslint --quiet` to something like:

```
eslint --quiet -c component-library.eslint.js
```

Where `component-library.eslint.js` is [a custom configuration file](https://eslint.org/docs/user-guide/configuring#configuration-file-formats) that imports the existing `.eslintrc` and exports it as json with our new rule added along with the export.

---

> With this process, if an engineer makes any changes to a file that imports the legacy component, they'll be forced to update the component usage before committing (or ignore the commit hook linting), right?

I believe so - that is the limitation of our current pre-commit linting. It lints entire changed files, not just the changed chunks.

---

> How might we reduce the burden for engineers making unrelated changes? Is there any way to catch only new imports for legacy components?

I mean... yeah, it's a solvable problem. We could capture the output from the `eslint` command, and if there is an error we could:

1. See if it is related to a `component-library` import
1. Check the git log or blame and see if the commit is before or after `HEAD`

Do you think this is something we need to put the time into?

cvalarida replied:
Cool. Let's work with the FE tools team on this when the time comes. We can iterate on the linting even after this doc gets approved.

cc @Clark-W-Griswold @meganhkelley for visibility

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+Once a new release is made for the Web Component library, we will [configure a linting rule](https://github.com/AlexMost/eslint-plugin-deprecate/blob/master/docs/rules/import.md) as part of the `lint:js:changed` script in `vets-website` to throw an error if a React component is imported that has a released Web Component replacement. This will prevent the older React component from appearing in mo ...
+
+Next, the Design System team will reach out to a VFS team that we believe will get the gratest value by adopting the Component. We will briefly describe the new Component and how it can be used in place of the older components currently used in their codebase, and we will offer to set up a meeting to synchronously go over the Component and begin the migration. The focus will be on making the migr ...
```
```suggestion
Next, the Design System team will reach out to a VFS team that we believe will get the greatest value by adopting the Component. We will briefly describe the new Component and how it can be used in place of the older components currently used in their codebase, and we will offer to set up a meeting to synchronously go over the Component and begin the migration. The focus will be on making the migration as smooth as possible for the team as well as bringing awareness to the new Component.
```

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+Next, the Design System team will reach out to a VFS team that we believe will get the gratest value by adopting the Component. We will briefly describe the new Component and how it can be used in place of the older components currently used in their codebase, and we will offer to set up a meeting to synchronously go over the Component and begin the migration. The focus will be on making the migr ...
+
+When we have finished helping the team transition all of their code to use the new Component, we will announce this publicly in Slack, and move on to help a new team with the migration, repeating the steps. After the new Web Component has replaced all instances of its React/`formation`/liquid conterparts, the React component will be removed from `component-library` and a new version will be publi ...
```
This approach will take a _lot_ of our very limited bandwidth. What I'd recommend instead is working with the first team directly _only_ if we believe the migration might be problematic. After that, we make the announcement and offer our normal support for any teams who need help making the migration.

The trouble with that approach, of course, is that we're not guaranteeing the sunsetting of legacy React components by directly being involved in their replacement.

The tradeoff here is our bandwidth vs having the legacy component stay in the codebase. For now, I'd say our time is more important to protect. We have other things that we need to be working on. I think the answer is that we won't provide support of the legacy components, but we will provide support for migrating to the new component if it's required. Down the road, we can re-evaluate based on the priority of this work compared to others, the time required, and whatever bandwidth we have at the time. (I'm still hoping for more team members. :sweat_smile: )

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+The code will live in the `component-library` repo on a dedicated orphan branch named `web-components`. This code will be used in `vets-website` via a dependency in `package.json`
+
+### Testing Plan
```
How will this be tested in `vets-website`? Do we need to create any guidance? Tools?

bkjohnson replied:
Some tests may need to be updated, i.e. the ones that are looking at the DOM for things specific to the React components. Some `skin-deep` tests use shallow rendering and may have a test that expects a `CollapsiblePanel` node to be rendered, for example. Other components may have specific classes that a test might look for, but with the Web Component implementation these classes may no longer be there or be within the shadow DOM.

I don't think we need to provide any tools specific to `vets-website`. If we find that there are common problems with a non-obvious solution related to testing, then anytime we make an announcement about a new WC I think we should include a blurb about that, like:

> Q: I'm getting [SPECIFC TEST ERROR], how do I fix it?
> 
> A: This has happened with other teams/components. We recommend XYZ.

cvalarida replied:
I guess what I'm asking about is how web components will interact with the existing e2e and unit tests. Like, how will they get loaded?

So I know we'll need to change those `skin-deep` tests, but when we interact with the web component in the unit test—like clicking on the `<va-accordion>` to ensure that the expected content is there—how will the functionality be there? We're using JSDOM and—presumably—aren't adding the WC loader to it to register the components and provide functionality. Unless I'm wrong? (I know basically nothing about testing web components.)

The tool I was asking about is more along the lines of providing the web component functionality in downstream (`vets-website`) unit tests which have to interact with the components.

If I'm still not making sense (because words are hard :sweat_smile:), please reach out in Slack and we can talk there. :+1:

cvalarida reviewed (commented):
I still have concerns about the migration, but offered a path forward. 30+ components * 20+ teams (no idea how accurate that is) is a _looooooooot_ of time spent helping folks migrate.

I understand not every team uses every component, and not every component is going to need a lot of help to migrate to, but committing to this is pretty aggressive.

bkjohnson committed:
Apply suggestions from code review

Co-authored-by: Chris Valarida <chris.valarida@adhocteam.us>

bkjohnson committed:
Mention the USWDS experiment

bkjohnson committed:
Update migration plan

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+**Author(s):** Brooks Johnson  
+**Last Updated:** January 13, 2021  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** _Chris Valarida_ \[ \], _Michael Fleet_ \[ \]
```
```suggestion
**Approvers:** _Chris Valarida_ [x], _Michael Fleet_ [ ]
```

cvalarida reviewed (approved):
Looks great! There are some solvable problems that we still have to work through:
- Linting rules (or something else entirely :shrug: ) to prevent new imports of deprecated components
- How to load the web components in `vets-website` unit tests
  - This is maybe already solved, but if so, we'll need to have the solution in writing here for posterity

All that said, I'm giving this the :+1: because I believe they're solvable. We'll just need to make sure to keep this document up-to-date if we make any changes down the road.

bkjohnson committed:
Add Chris's approval

Co-authored-by: Chris Valarida <chris.valarida@adhocteam.us>

bkjohnson committed:
Add Dror to list of approvers

cvalarida commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+How will we measure performance of Web Components, lazy-loading, etc?
+
+#### Risks
```
Another good point @drorva mentioned to me is industry adoption. My understanding is it's been pretty slow going heretofore, so if the standard ends up becoming obsolete, that may affect the future of this work. I don't know what the risk of that is, but it's worth calling out.

bkjohnson replied:
Included that point in this commit: https://github.com/department-of-veterans-affairs/va.gov-team/pull/18662/commits/86772e69b3b2d04821684374714f1df6c16135d6

f1337 commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+### Objective
+
+The goal is to establish a component library using the [Web Components standard](https://developer.mozilla.org/en-US/docs/Web/Web_Components). This will eventually supercede the current `component-library` code and some of the liquid HTML templates currently in `vets-website`. A non-goal is creating an entirely new visual experience as part of the component library.
```
Why do we want a Web Components library? Please add one sentence in this paragraph, probably right after the first, summarizing the value of web components vs our existing approach.

f1337 commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,221 @@
+
+### High Level Design
+
+The goal is to establish a single codebase and documentation for the implementation of design system components. This will allow the components to be used in React apps, static HTML, or any other web framework.
```
This sentence is probably what I'm asking for in the Objective paragraph. Consider moving it up?

f1337 reviewed (commented):
I made two inline comments re: clarifying the Why? in the opening Objective section.

I'd also like to see some example code, maybe at the end of the doc, addressing the following:
* uber simple Web Component
* using WC in React
* using WC in plain ole HTML

bkjohnson committed:
Add "why" to the Objective section

bkjohnson committed:
Merge branch 'wc-design-doc' of github.com:department-of-veterans-affairs/va.gov-team into wc-design-doc

bkjohnson committed:
Add simple examples of a Web Component in use

bkjohnson committed:
Remove images & show diff instead

bkjohnson committed:
Add a newline to hopefully fix formatting

bkjohnson commented:
I've added the "why" to the Objective section, and added some example code as well. I was going to include images, but I couldn't get them to resize properly, and including the rendered page along with the inspector made for a large and awkwardly placed screenshot in the document.

bkjohnson committed:
Mention risk of obsolescence

bkjohnson commented:
@drorva and @f1337, do you have any thoughts on the changes?

bkjohnson committed:
Update testing section

f1337 reviewed (approved):
Excellent! :metal:
![Bill & Ted playing air guitar, saying 'Excellent!'](https://media.giphy.com/media/l46CDHTqbmnGZyxKo/giphy.gif)

bkjohnson committed:
Update approvers

I accidentally removed Chris's approval as a result of a merge combined with an older local branch

bkjohnson committed:
Call out uswds dependency in formation

bkjohnson commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,319 @@
+
+The intended audience for this document is frontend engineers.
+
+### Background
+
+Currently, the design system has various implementations of "components" scattered in various places to serve various tools/libraries:
+
+- The `formation` npm package
+    - Built using `uswds` 1.6
+    - This contains static HTML, CSS, and vanilla JavaScript
+    - Used by copy-pasting markup onto static pages
+-  The `component-library` package
+    - This is the rebranded `formation-react`, containing React implementations of some patterns defined in `formation`
+    - Can only be used inside React apps
+- [Liquid templates in `vets-website`](https://github.com/department-of-veterans-affairs/vets-website/tree/958d89f18db29a1f3e1f1cbe4e304299eb387de7/src/site)
+    - This is similar to the static markup found in `formation`
+    - These templates of components are only used in the content build for va.gov
```
@drorva , does this answer your question:

> I would add a section in the beginning of the doc that describes the current state. As I understand it, we forked uswds 1.0, use react components (based on some library?), etc. This provides context.

I recently added the bullet point:

`     - Built using `uswds` 1.6 `

bkjohnson commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,319 @@
+
+Next, if we believe a particular component migration will be problematic the Design System team will reach out to a VFS team to work with them on the migration. Once the first team has completed their migration to the new Component, we will announce the success and encourage other teams to do the same. At this point we will provide support for the migration if requested, but we will no longer pro ...
+
+After the new Web Component has replaced all instances of its React/`formation`/liquid conterparts, the React component will be removed from `component-library` and a new version will be published. The associated styles will also be removed from `formation` if we are certain that this will not affect other styles on va.gov.
```
Does this answer the question about the frontend?

> Similar to the previous point, what does the full front-end stack look like with Web components on va.gov? Do developers still need to use Formation, react, etc. It's fine to point to info on https://design.va.gov/.

bkjohnson replied:
I added a bit more detail to this: https://github.com/department-of-veterans-affairs/va.gov-team/pull/18662/commits/ce991f7eb3a463e1456c6e4bbaee8e06d0aef957

bkjohnson committed:
Confirm browser support w/ polyfills

bkjohnson commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,319 @@
+
+#### Loading Web Components on va.gov
+
+The setup code for the WC library will be in its own `vets-website` entry file that is loaded on each page. This code will be responsible for [applying polyfills](https://stenciljs.com/docs/faq#what-polyfills-does-stencil-provide) for older browsers as well as [defining the Web Components](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define). We will be able to achieve [ ...
```
Does this answer your question? I just added the last sentence beginning with "We will be able to achieve ..."

> https://developer.mozilla.org/en-US/docs/Web/Web_Components#browser_compatibility indicates limits in Edge and Safari beyond the ones that we're aware of in IE. Are we comfortable that browser shims solve for the feature set we need?

bkjohnson committed:
Organize risks & add one for abandoned development

bkjohnson commented on platform/engineering/design-docs/2021-01-13-web-component-library.md:
```diff
@@ -0,0 +1,337 @@
+
+An additional risk would be if certain features of the Web Component standard are deprecated or otherwise become obselete. We are already in a similar situation with React components - our `component-library` consists of many "legacy" class-based components rather than functional components which have the option to use hooks.
+
+##### Abandoned development
+
+Web Components are a feature that is already built into modern browsers, but they don't have as wide of an industry adoption as a framework like React. Despite this, it is unlikely that the Web Component standard will be completely abandoned. There are many large companies which are already developing products using Web Components:
+
+- Youtube
+- GitHub
+- Xfinity (Comcast)
+- [Apple Music](https://hackernoon.com/apple-cements-the-unlikely-rise-of-web-components-6b1d3g1t)
+- [Adobe](https://opensource.adobe.com/spectrum-web-components/)
+
+There aren't any published articles detailing Web Component adoption by the companies without links, but Web Components can be found on their websites.
```
This is a new section added in response to your comment:

> While I, like many other open source and gov people, like the idea of platform Web Components, I am not clear that I see industry adoption of this approach. I worry that down the line we'll need to pivot again because this will be a dead-end. What would reassure me is finding out that some major organization are using it.

bkjohnson reviewed (commented):
@drorva , I've made some changes in response to your questions/comments in Slack.

bkjohnson committed:
Add a bit more detail regarding full frontend stack

f1337 reviewed (approved):
AFAIC, we're good to move forward. I'm very excited to see where the experiment leads!

![More Cowbell](https://media.giphy.com/media/3o7TKnqKj5vE0kC0ik/giphy.gif)

bkjohnson committed:
Update before merging
